streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -990,6 +991,10 @@ public void unregister(final Collection<TopicPartition> revokedChangelogs) {
         }
 
         removeChangelogsFromRestoreConsumer(revokedInitializedChangelogs);
+
+        if (changelogs.isEmpty()) {
+            state = ChangelogReaderState.ACTIVE_RESTORING;
```

Comment:
Apologies if I'm missing something or my understanding of the restoration-related code is just completely out of date now, but do I understand correctly that this was/is a bug in the current (soon-to-be-old) architecture, and not related to the restoration thread work? 

Also, just curious, did you find this bug due to you or someone else hitting it, or was it just something you happened to notice wasn't quite right? If you/someone is actually running into this "in the wild" it may be a good idea to file a jira ticket for it, just to document that the bug exists, we're aware of it, and that it's fixed in whichever versions.

Reply:
Yeah, this change is maybe a bit confusing without further context because you only see the fix, not really where the bug is happening. 

This fixes a bug that is only triggered in the new code path. So there is no risk that somebody in the wild will run into it. Basically, the bug is triggered if you `transitToUpdateStandby` before making sure that your `ChangeLogReader` is in `ACTIVE_RESTORING` mode. The old code path just calls `enforceRestoreActive` no matter what, while the new code path doesn't - if you happen to remove all tasks and then insert a standby task. I considered also calling `enforceRestoreActive` in the new path, just in case, but I think it is cleaner to have a well-defined invariant for this class (which is `changelogs.isEmpty` implies `state=ACTIVE_RESTORING`).